puppet.go

```diff
@@ -22,68 +22,119 @@ import (
+	// as soon as the YAML data overflows the buffer.
+	data := make([]byte, PUPPET_YAML_BUFFER_SIZE)
 	var count int
 	count, err = stdout.Read(data)
 	if err != nil && err != io.EOF {
-		log.Printf("Error reading catalog data from puppet: %v", err)
-		return nil
+		log.Printf("Puppet: Error reading YAML data from puppet: %v", err)
+		return nil, err
 	} else if DEBUG {
-		log.Printf("read %v bytes of data from puppet", count)
+		log.Printf("Puppet: read %v bytes of data from puppet", count)
 	}
 	for scanner := bufio.NewScanner(stderr); scanner.Scan(); {
-		log.Printf("Errors[puppet]: %v", scanner.Text())
+		log.Printf("Puppet: (output) %v", scanner.Text())
 	}
 	if err := cmd.Wait(); err != nil {
-		log.Printf("Error: puppet command did not complete: %v", err)
-		return nil
+		log.Printf("Puppet: Error: puppet command did not complete: %v", err)
+		return nil, err
 	}
+
 	// Slicing down to the number of actual bytes is important, the YAML parser
 	// will choke on an oversized buffer. http://stackoverflow.com/a/33726617/3356612
-	if err := config.Parse(data[0:count]); err != nil {
-		log.Printf("Error: Config: ParseConfigFromFile: Parse: %v", err)
+	return data[0:count], nil
+}
+
+func ParseConfigFromPuppet(puppetParam, puppetConf string) *GraphConfig {
+	var puppetConfArg string
+	if puppetConf != "" {
+		puppetConfArg = "--config=" + puppetConf
+	}
+
+	var cmd *exec.Cmd
+	if puppetParam == "agent" {
+		cmd = exec.Command("puppet", "mgmtgraph", "print", puppetConfArg)
+	} else if strings.HasSuffix(puppetParam, ".pp") {
+		cmd = exec.Command("puppet", "mgmtgraph", "print", puppetConfArg, "--manifest", puppetParam)
+	} else {
+		cmd = exec.Command("puppet", "mgmtgraph", "print", puppetConfArg, "--code", puppetParam)
+	}
+
+	log.Println("Puppet: launching translator")
+
+	var config GraphConfig
+	if data, err := runPuppetCommand(cmd); err != nil {
+		return nil
+	} else if err := config.Parse(data); err != nil {
+		log.Printf("Puppet: Error: Could not parse YAML output with Parse: %v", err)
 		return nil
 	}
 
 	return &config
 }
+
+func PuppetInterval(puppetConf string) chan bool {
+	if DEBUG {
+		log.Printf("Puppet: determining graph refresh interval")
+	}
+	var cmd *exec.Cmd
+	if puppetConf != "" {
+		cmd = exec.Command("puppet", "config", "print", "runinterval", "--config", puppetConf)
+	} else {
+		cmd = exec.Command("puppet", "config", "print", "runinterval")
+	}
+
+	log.Println("Puppet: inspecting runinterval configuration")
+
+	interval := 1800
+	if data, err := runPuppetCommand(cmd); err != nil {
+		log.Printf("Puppet: could not determine configured run interval (%v), using default of %v", err, interval)
+	} else {
+		result, err := strconv.ParseInt(strings.TrimSpace(string(data)), 10, 0)
+		if err != nil {
+			log.Printf("Puppet: error reading numeric runinterval value (%v), using default of %v", err, interval)
+		} else {
+			interval = int(result)
+		}
+	}
+
+	ch := make(chan bool)
```

Comment:
It won't compile though. I tried to go for `Tick` rather than `After` (we do want regular events), but it's a type mismatch.

```
./puppet.go:137: cannot use time.Tick(time.Duration(interval) * time.Second) (type <-chan time.Time) as type chan bool in return argument
```